joomla/joomla-cms
# [4.0] Session Manager for destroying sessions, fix session destruction in user privacy plugin

Partial Pull Request for Issue #26539

### Summary of Changes

Force logout of a user does not work consistently unless you are using the database as your session handler.  This pull request for 4.0 refactors the system to ensure all functions work with all handlers.

The session handler used internally by `Joomla\CMS\Session\Session` is now exposed as a service, this should never be direct accessed but is needed to make the new manager class work.

The `Joomla\CMS\Session\SessionManager` class is introduced as a means for allowing someone to destroy a session's data using the session handler.

The "User - Joomla!" plugin is adjusted to use the new session manager to handle session destruction when deleting a user or when the force logout option is enabled at logout (contingent on session metadata being available).

The "Privacy - User" plugin is adjusted to do the same when removing data (currently this will fatal out because the plugin tries using the no longer existing `JSessionStorage` class).

### Testing Instructions

- Ensure you can log in correctly
- Ensure you can log out correctly
- With session metadata enabled (default) and the "Force Logout for all Sessions?" option enabled (default) in the "User - Joomla!" plugin, all sessions should be destroyed when logging out regardless of session handler (should be validated using at least database and filesystem, if you have others available then bonus points to you)
- With session metadata disabled and the "Force Logout for all Sessions?" option enabled (default) in the "User - Joomla!" plugin, only the current session should be destroyed when logging out
- With session metadata enabled (default) and the "Force Logout for all Sessions?" option enabled (default) in the "User - Joomla!" plugin, logging a user out from the "Logged-in Users" module should log out all sessions for the given user
- With session metadata enabled (default) and the "Force Logout for all Sessions?" option disabled in the "User - Joomla!" plugin, logging a user out from the "Logged-in Users" module will not work
    - Fixing this is beyond scope of this pull request, the ability to destroy all sessions for a user has always been reliant on this option being enabled and to make it work with the option disabled will require some way of passing a signal from the logout controller to the application and the plugin(s) handling the logout event
- With session metadata disabled, you shouldn't even see a user list in the "Logged-in Users" module to do anything with the account

### Expected result

- Logging a user out with the icon in the "Logged-in Users" module logs a user out regardless of session handler
- Removing a user account with a privacy request does not trigger a fatal error

### Actual result

- Logging a user out with the icon in the "Logged-in Users" module logs a user out only when using the database session handler
- Removing a user account with a privacy request triggers a fatal error due to the use of a class removed from 4.0

### Documentation Changes Required

If the DI container services are being documented, make it clear as day that nobody should EVER, EVER, EVER direct access the 'session.handler' service (or any of its aliases).  Introducing the concept of private services to the container so that this service is only accessible in the scope of a service factory would be a good way of preventing this, but if not added to the container before 4.0 then the feature is useless until 5.0 because if introduced, moving a service from public to private should be considered a B/C break (in the same way moving the visibility of a class method or variable from public to protected or private would be one).
